src/fido2_child/fido2_child_common.c

```diff
@@ -0,0 +1,313 @@
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <popt.h>
+#include <sys/prctl.h>
+#include <fido/param.h>
+
+#include "util/debug.h"
+#include "util/util.h"
+
+#include "fido2_child.h"
+
+errno_t
+cose_str_to_int(const char *type, int *out)
+{
+    if (strcasecmp(type, "es256") == 0) {
+        *out = COSE_ES256;
+    } else if (strcasecmp(type, "rs256") == 0) {
+        *out = COSE_RS256;
+    } else if (strcasecmp(type, "eddsa") == 0) {
+        *out = COSE_EDDSA;
+    } else {
+        *out = 0;
+        return ERR_INVALID_CRED_TYPE;
+    }
+
+    return EOK;
+}
+
+errno_t
+parse_arguments(int argc, const char *argv[], struct fido2_data *data)
+{
+    int opt;
+    int dumpable = 1;
+    int debug_fd = -1;
+    char *user_verification = NULL;
+    const char *opt_logger = NULL;
+    const char *type = NULL;
+    poptContext pc;
+    errno_t ret;
+
+    // Set defaults
+    data->action = ACTION_NONE;
+    data->shortname = NULL;
+    data->domain = NULL;
+    data->public_key = NULL;
+    data->key_handle = NULL;
+    data->type = COSE_ES256;
+    data->user_verification = FIDO_OPT_OMIT;
+    data->request_pin = false;
+    data->debug_libfido2 = false;
+
+    struct poptOption long_options[] = {
+        POPT_AUTOHELP
+        SSSD_DEBUG_OPTS
+        {"dumpable", 0, POPT_ARG_INT, &dumpable, 0,
+         _("Allow core dumps"), NULL },
+        {"debug-fd", 0, POPT_ARG_INT, &debug_fd, 0,
+         _("An open file descriptor for the debug logs"), NULL},
+        SSSD_LOGGER_OPTS
+        {"register", 0, POPT_ARG_NONE, NULL, 'r',
+         _("Register a FIDO2 key for a user"), NULL },
+        {"authenticate", 0, POPT_ARG_NONE, NULL, 'a',
+         _("Authenticate a user with a FIDO2 key"), NULL },
+        {"username", 0, POPT_ARG_STRING, &data->shortname, 0,
+         _("Shortname"), NULL },
+        {"domain", 0, POPT_ARG_STRING, &data->domain, 0,
+         _("Domain"), NULL},
+        {"public-key", 0, POPT_ARG_STRING, &data->public_key, 0,
+         _("Shortname"), NULL },
+        {"key-handle", 0, POPT_ARG_STRING, &data->key_handle, 0,
+         _("Domain"), NULL},
+        {"type", 0, POPT_ARG_STRING, &type, 0,
+         _("COSE type to use"), NULL},
+        {"user-verification", 0, POPT_ARG_STRING, &user_verification, 0,
+         _("Require user-verification"), NULL},
+        {"pin", 0, POPT_ARG_NONE, NULL, 'p',
+         _("Require PIN"), NULL },
```

Comment:
That was my initial intention but after reading the FIDO2 specification regarding the [built-in User Verification method](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#built-in-user-verification-method) I changed my mind.

The specification doesn't accept the PIN as a user verification method, which leads us to a separate management of them (i.e. `fido_dev_has_pin()`, `fido_dev_has_uv()`). We could use just one parameter as an argument to enable the user verification (both PIN and built-in user verification) from the organization side, but what would happen if none of them is enabled in the key side? Which of the two should be enabled? How can the user decide which one is better?

By the way, I still have my doubts regarding this decision so any feedback is much welcome.

Reply:
To be clear, the specification refers to _either_ `clientPin` or built-in user verification as "[some form of user verification](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#some-form-of-user-verification)". The section that you linked clarifies that while `clientPin` is a method for user verification it is not _built-in_ to the authenticator (the authenticator does not collect the PIN itself, the client does). Both methods will make the authenticator set the "uv" flag in the response.

The CTAP specification [_suggests_ (non-normative)](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#sctn-makeCred-platf-actions) that if the Relying Party wants to perform user verification or if the authenticator will require it, built-in UV is to be preferred if that is supported by the authenticator. If built-in UV fails or is unsupported,  `clientPin` can be used as a fallback if that is supported by the authenticator. 

As a further example, see the WebAuthn specification where Relying Parties can set only set (_some form of_) user verification to ["discouraged", "preferred", or "required"](https://www.w3.org/TR/webauthn-2/#enum-userVerificationRequirement). The browser then turns that into the appropriate request for the user's authenticator.

A very simplified/incomplete libfido2 example could look something like (with error handling omitted):
```c
bool has_uv = fido_dev_has_uv(dev);
bool has_pin = fido_dev_has_pin(dev);
char *pin = NULL;

if (user_verification_is_required && !has_uv && !has_pin) {
	// TODO: Help the user enroll a PIN/fingerprint or similar, if possible.
	return -1; 
}
if (has_uv) {
	fido_cred_set_uv(cred, FIDO_OPT_TRUE);
} else if (has_pin) {
	read_pin(&pin);
}
r = fido_dev_make_cred(dev, cred, pin);
if (r == FIDO_ERR_PIN_REQUIRED) {
	read_pin(&pin);
	r = fido_dev_make_cred(dev, cred, pin);
}
```

Again, this depends mostly on your preferred UX. Since you're implementing both the platform/client and the "API" for your Relying Parties, you have a bit more flexibility to choose from. :slightly_smiling_face:

Reply:
Understood. I've removed the pin option and I have included part of your proposals. Please take a look at them in [generate_credentials()](https://github.com/SSSD/sssd/pull/6327/files#diff-feaa53c7add9db70839332290a3279a54edcf0e319b85b3525e111ad0a64b638R151).